### PR TITLE
RavenDB-23158 Cache the table instances (and the associated new page …

### PR DESCRIPTION
**Pull request by ayende**
…allocator) at the context level

### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-23158 

### Additional description

Caching a table instance for the lifetime of a transaction instead of creating a new one each time.

### Type of change

- [x] Optimization

### How risky is the change?

- [x] Low 

### Backward compatibility
- [x] Not relevant

### CODE REVIEW

File: src/Raven.Server/Documents/CountersStorage.Sharding.cs [hunk 1]
```diff
@@ -25,7 +25,7 @@ public IEnumerable<ReplicationBatchItem> GetCountersByBucketFrom(DocumentsOperat
 
         public IEnumerable<CounterTombstoneDetail> GetCounterTombstonesByBucketFrom(DocumentsOperationContext context, int bucket, long etag)
         {
-            var table = new Table(CounterTombstonesSchema, context.Transaction.InnerTransaction);
+            var table = context.CountersTable(this);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/CountersStorage.Sharding.cs]: shouldn't this be CounterTombstones?

File: src/Raven.Server/Documents/CountersStorage.cs [hunk 2]
```diff
@@ -1873,7 +1873,7 @@ private static ChangeVector ExtractCounterTombstoneChangeVector(DocumentsOperati
 
         public IEnumerable<CounterTombstoneDetail> GetCounterTombstonesFrom(DocumentsOperationContext context, long etag, long toEtag = long.MaxValue)
         {
-            var table = new Table(CounterTombstonesSchema, context.Transaction.InnerTransaction);
+            var table = context.CountersTable(this);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/CountersStorage.cs]: shouldn't this be CounterTombstones?

**Review by ppekrol — COMMENTED** [review]
Do you think we need that as well for Indexes and Cluster?

File: src/Raven.Server/Documents/CountersStorage.Sharding.cs (hunk 1 above)
Comment (redknightlois) [reply src/Raven.Server/Documents/CountersStorage.Sharding.cs]: Yes.

File: src/Raven.Server/Documents/CountersStorage.cs (hunk 2 above)
Comment (redknightlois) [reply src/Raven.Server/Documents/CountersStorage.cs]: Yes

**Comment by ayende** [comment]
I changed that _everywhere_. That was the only place in which we used naked `new Table(` like this

File: src/Raven.Server/Documents/CountersStorage.Sharding.cs (hunk 1 above)
Comment (ayende) [reply src/Raven.Server/Documents/CountersStorage.Sharding.cs]: Fixed

File: src/Raven.Server/Documents/CountersStorage.cs (hunk 2 above)
Comment (ayende) [reply src/Raven.Server/Documents/CountersStorage.cs]: Fixed

**Comment by ppekrol** [comment]
Rebased and fixed conflicts

**Pull request merged**